### ISSUE TITLE
[HOLD for payment 2022-03-17] [HOLD for payment 2022-03-09] Feature Request: Add image preview placeholder while uploading images/Videos - Reported by @mdneyazahmad

### ISSUE THREAD

**Issue by mvtglobally** [Expensify/App#7463]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app
2. Go to chat on slower network 
3. Upload image and note the loading 

## Expected Result:
Image preview should display while uploading images

## Actual Result:
No Image preview displayed while uploading images

## Workaround:
unknown 

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.33-0 
**Reproducible in staging?:** Y 
**Reproducible in production?:** Y 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation 
![Screenshot_2022-01-19-00-00-25-84_4f9154176b47c00da84e32064abf1c48](https://user-images.githubusercontent.com/43995119/151732467-050436b4-46de-475d-af8c-b21edcc0e79a.jpg)

**Expensify/Expensify Issue URL:**
**Issue reported by:** @mdneyazahmad
**Slack conversation:**  https://expensify.slack.com/archives/C01GTK53T8Q/p1642530762139600

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @JmillsExpensify (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by JmillsExpensify** [comment]
I agree that from a best practice perspective, we should be providing a local preview of the receipt on upload. That said, I think we want to be careful on how we implement this, so we'll want to make sure any proposals are discussed here and in Slack. I'll go ahead and create a job in the meantime though!

**Comment by JmillsExpensify** [comment]
Job is here: https://www.upwork.com/jobs/~01e1ba3f717ab3ff5d

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Julesssss (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by wahaj-47** [comment]
I would use the temporary local image path and pass it to the Image component. We can continue using the local image path as long as the user stays on this screen, and the next time the user opens this screen the image (from a server url) should load with the rest of the messages.

**Comment by parasharrajat** [comment]
Thanks, @wahaj-47  for the proposal. Could you please elaborate on your proposal?
1. Where will you make changes in the codebase? 
2. Using our codebase, how will you achieve this?
3. Break down your proposal in steps.
Do you have any questions?

**Comment by wahaj-47** [comment]
I am looking at the code, and in case of an attachment, while its uploading an ActivityIndicator is rendered but once the upload is complete, is the image being rendered as HTML?

https://github.com/Expensify/App/blob/2c14f01af550f554bc1f488a35654c36b437ba65/src/pages/home/report/ReportActionItemFragment.js#L49-L86

**Comment by okansahin** [comment]
# Proposal

First I add picked file info to REPORT_ACTIONS collection to using local file path in ReportedActionItemFragment component.

https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/libs/actions/Report.js#L1098-L1099

```    
const attachmentInfo = isAttachment ? file : {};
```

https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/libs/actions/Report.js#L1162-L1164
```
  isAttachment,
  attachmentInfo,
  loading: true,
```

And pass this prop to ReportedActionItemFragment from ReportedActionItemMessage. 

https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/pages/home/report/ReportActionItemMessage.js#L42-L47

```
<ReportActionItemFragment
                            key={`actionFragment-${props.action.sequenceNumber}-${index}`}
                            fragment={fragment}
                            isAttachment={props.action.isAttachment}
                            attachmentInfo={props.action.attachmentInfo}
                            loading={props.action.loading}
                    />
```

Crate proptype and default prop. Check the attachment is image and show with ReactNative's ImageBackground.

https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/pages/home/report/ReportActionItemFragment.js#L25-L29

```
    /** Is this fragment an attachment? */
    isAttachment: PropTypes.bool,

    /** If this fragment is attachment than has info? */
    attachmentInfo: PropTypes.shape({
        name: PropTypes.string,
        size: PropTypes.number,
        type: PropTypes.string,
        uri: PropTypes.string,
    }),

    /** Does this fragment belong to a reportAction that has not yet loaded? */
    loading: PropTypes.bool,
```

https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/pages/home/report/ReportActionItemFragment.js#L40-L45

```
const defaultProps = {
    isAttachment: false,
    attachmentInfo: {
        name: '',
        size: 0,
        type: '',
        uri: '',
    },
    loading: false,
    isSingleLine: false,
    tooltipText: '',
};
```

https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/pages/home/report/ReportActionItemFragment.js#L51-L61

```
      if (props.isAttachment && props.loading) {
          return (
              <View style={[styles.chatItemAttachmentPlaceholder]}>
                  {Str.isImage(props.attachmentInfo.uri)
                      ? (
                          <ImageBackground
                              source={{uri: props.attachmentInfo.uri}}
                              resizeMode="cover"
                              imageStyle={[styles.borderBottomRounded, styles.borderTopRounded]}
                              style={[styles.flex1, styles.justifyContentCenter]}
                          >
                              <ActivityIndicator
                                  size="large"
                                  color={themeColors.textSupporting}
                                  style={[styles.flex1]}
                              />
                          </ImageBackground>
                      ) : (
                          <ActivityIndicator
                              size="large"
                              color={themeColors.textSupporting}
                              style={[styles.flex1]}
                          />
                      )}
              </View>
          );
      }
```

And maybe we can add a fade background to activity indicator because its hard to visible with some pictures.

**Comment by parasharrajat** [comment]
Thanks for the proposal @okansahin. 

Questions: 
1. > First I add picked file info to REPORT_ACTIONS collection to using local file path in ReportedActionItemFragment component.

Will this work on all platforms? Is a local path is available on the web? 

2. > And maybe we can add a fade background to activity indicator because its hard to visible with some pictures. 

That would be great,

**Comment by wahaj-47** [comment]
I was thinking instead of using the local path directly like I proposed we use the same approach used in the AttachementModal to display a preview. It should work on all platforms.

**Comment by okansahin** [comment]
Thank you @parasharrajat 

> Thanks for the proposal @okansahin.
> 
> Questions:
> 
> 1. > First I add picked file info to REPORT_ACTIONS collection to using local file path in ReportedActionItemFragment component.
> 
> Will this work on all platforms? Is a local path is available on the web?

No, its because Str.isImage check the extension of url, but web returns file uri something like blob:http://localhost:8080/2d9a77b1-8647-48ae-8994-1504a449c521

But without Str.isImage is okay on desktop, android, ios, web including safari, chrome, firefox.

I can handle it with any other condition.

> 
> 2. > And maybe we can add a fade background to activity indicator because its hard to visible with some pictures.
> 
> That would be great,

I will add a faded rounded square background to it.

**Comment by parasharrajat** [comment]
@wahaj-47 There are a few questions for you https://github.com/Expensify/App/issues/7463#issuecomment-1026632472. 

They will help me understand your proposal better. Let me know if you are planning to submit one. I can wait.

**Comment by parasharrajat** [comment]
> I can handle it with any other condition.

Maybe try checking the name instead.

**Comment by okansahin** [comment]
@parasharrajat 

Yes, you completely right. I follow the next comments to have finish the task.

**Comment by okansahin** [comment]
> I was thinking instead of using the local path directly like I proposed we use the same approach used in the AttachementModal to display a preview. It should work on all platforms.

@wahaj-47 Can you explain more about "use the same approach used in the AttachementModal"?

**Comment by parasharrajat** [comment]
@okansahin 's [proposal](https://github.com/Expensify/App/issues/7463#issuecomment-1027963336) looks good to me.  

cc: @Julesssss 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by wahaj-47** [comment]
### Proposal
We add the local uri in the REPORT_ACTION collection
https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/libs/actions/Report.js#L1098
```
let localUri = null;

if (isAttachment) {
  if (file instanceof File) {
    localUri = URL.createObjectURL(file);
  } else {
    localUri = file.uri;
  }
}
```
https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/libs/actions/Report.js#L1163
```
isAttachment,
localUri,
```

Pass this prop to ReportedActionItemFragment.
https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/pages/home/report/ReportActionItemMessage.js#L42-L47
```
<ReportActionItemFragment
        key={`actionFragment-${props.action.sequenceNumber}-${index}`}
        fragment={fragment}
        isAttachment={props.action.isAttachment}
        loading={props.action.loading}
        localUri={props.action.localUri}
/>
```

And use the AttachmentView component to display the attachment
https://github.com/Expensify/App/blob/ad87352530e8d3a73860a1921af07f9288aad87f/src/pages/home/report/ReportActionItemFragment.js#L51-L61
```
if (props.isAttachment && props.loading) {
    return (
        <View style={[styles.chatItemAttachmentPlaceholder]}>
            <AttachmentView sourceUrl={props.localUri}></AttachmentView>
            <ActivityIndicator
                size="large"
                color={themeColors.textSupporting}
                style={[styles.flex1]}
            />
        </View>
    );
}
```

**Comment by okansahin** [comment]
> 

Thank you @parasharrajat 

Then I send proposal on Upwork. Thank you.

**Comment by MelvinBot** [comment]
📣 @okansahin You have been assigned to this job by @Julesssss!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
Current assignee @JmillsExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by Julesssss** [comment]
@okansahin I'm holding the assignment until the UpWork job is created and assigned, it shouldn't take long

**Comment by MelvinBot** [comment]
📣 @okansahin You have been assigned to this job by @Julesssss!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by Julesssss** [comment]
Whoops, I blame Melvin as it didn't immediately comment and I thought we weren't yet ready. Please go ahead @okansahin 😄

**Comment by okansahin** [comment]
> Job is here: https://www.upwork.com/jobs/~01e1ba3f717ab3ff5d

@Julesssss here is a job post that I sent a proposal. Will you open a new?

**Comment by Julesssss** [comment]
Hi @okansahin, please ignore my other message, I didn't see that issue. It's the correct one 👍

**Comment by okansahin** [comment]
@Julesssss Okay. I sent and wait to be accepted my proposal in Upwork.

**Comment by JmillsExpensify** [comment]
Hired assigned contributors associated with this job.

**Comment by JmillsExpensify** [comment]
Active convos in the related PR. Check there for the latest.

**Comment by Julesssss** [comment]
Merged, awaiting deployment to prod

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.38-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/7569


If no regressions arise, payment will be issued on 2022-02-22. :confetti_ball:

**Comment by JmillsExpensify** [comment]
Woo! Now it's just a question of time. I'll circle back for payment following the regression period.

**Comment by rushatgabhane** [comment]
Heads up, we discovered a regression while testing https://github.com/Expensify/App/pull/7723 - Image preview placeholder doesn't appear when uploading the image by drag and drop.

https://user-images.githubusercontent.com/29673073/155135495-f4410d0d-be0f-40b2-ad74-2b1d83baa127.mp4



Thanks to @okansahin for connecting the dots and they already have a fix for it. https://github.com/okansahin/App/commit/63213b0922f85954f45ea6a944f902d04ab79432

Let's create a follow-up PR to fix the regression.

**Comment by okansahin** [comment]
> Heads up, we missed a case - Image preview placeholder doesn't appear when uploading the image by drag and drop.
> 
>  153815228-9acd7a4c-feac-4771-8930-705fd6271155.mp4 
> Thanks to @okansahin for connecting the dots in #7723 and they already have a fix for it. [okansahin@63213b0](https://github.com/okansahin/App/commit/63213b0922f85954f45ea6a944f902d04ab79432)
> 
> Let's create a follow-up PR to fix this.

I also added new prop and it's comment to PropTypes.

https://github.com/Expensify/App/issues/7463#issuecomment-1047763484

Do I create new follow-up PR?

**Comment by parasharrajat** [comment]
Thanks, @rushatgabhane. I didn't notice that during the PR or proposal review. Yeah @okansahin Please create a follow-up. I have seen your [changes](https://github.com/Expensify/App/compare/main...okansahin:okansahin/feature/upload-image-preview-placeholder). 

I would suggest using source property on file and pass it from https://github.com/Expensify/App/blob/58f0168373e022440a01a966a0bae772f01dda1e/src/components/AttachmentModal.js#L208.

**Comment by okansahin** [comment]
> I would suggest using source property on file and pass it from
> 
> https://github.com/Expensify/App/blob/58f0168373e022440a01a966a0bae772f01dda1e/src/components/AttachmentModal.js#L208
> 
> .

Thanks @parasharrajat for your advice. 

Is it okay to create new PR?

https://github.com/okansahin/App/commit/518f73416b498b859ce9d26e3be7f2c4638afeef

**Comment by parasharrajat** [comment]
Yup. Go ahead.

**Comment by JmillsExpensify** [comment]
Holding on payment until we get the most recently linked PR merged.

**Comment by JmillsExpensify** [comment]
It looks like we've had some issues with the most recent linked PR. Is that right, and are we actively working through those?

**Comment by parasharrajat** [comment]
There was no issue with the PR for this issue but later OP found another case that was missed from the first PR so he created another PR to cover that.

**Comment by Julesssss** [comment]
Hi @okansahin, we'll reset the regression timer to when your second PR was merged. So you should be due for payment on Wednesday 9th

**Comment by JmillsExpensify** [comment]
@mdneyazahmad I just noticed that you're not on the related Upwork job and we'd like to pay you for reporting. Can you please share your Upwork profile so I can make sure to invite you to the job? In the meantime, I'm going to pay out the other contributors now that the regression period has ended.

**Comment by mdneyazahmad** [comment]
@JmillsExpensify https://www.upwork.com/freelancers/~011922ae9ff280b160 Thank you

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.41-6 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/7874


If no regressions arise, payment will be issued on 2022-03-17. :confetti_ball:

**Comment by JmillsExpensify** [comment]
Ah interesting, it looks like Jules and I aligned payment to the first PR, not the second one.

**Comment by okansahin** [comment]
Ah, may it was my fault. I asked for a regression period from @Julesssss 

I thought it was production deploy but now I see it is staging. I'm sorry about that. I follow the issue to fix if any trouble.

@Julesssss I am really sorry to make confused you too.

**Comment by JmillsExpensify** [comment]
No worries! I'm going to go ahead and update the issue with the updated date so that we don't forget. I'll go ahead and close it out if we get through the regression period and no issues.

**Comment by JmillsExpensify** [comment]
Circling back to this issue to formally close us out.

**Comment by mdneyazahmad** [comment]
@JmillsExpensify I noticed that the job for this issue on Upwork is not closed. Thank you

**Comment by JmillsExpensify** [comment]
<img width="1060" alt="Screen Shot 2022-03-25 at 1 02 01 PM" src="https://user-images.githubusercontent.com/1127863/160185058-2a860956-ef11-486f-8277-eb8b468d441c.png">

Sorry can you help clarify?
